cmd/object-handlers.go

```diff
@@ -152,8 +176,72 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 		length = hrange.getLength()
 	}
 
+	getObject := objectAPI.GetObject
+	if api.CacheAPI() != nil && !hasSSECustomerHeader(r.Header) {
+		getObject = api.CacheAPI().GetObject
+	}
+
 	var writer io.Writer
-	writer = w
+
+	if objInfo.IsCompressed() {
+		var pipeErr error
+
+		// Defaulting the length and startOffset.
+		// Incase of range based queries, the entire data is read to the pipeWriter.
+		// So that range is set on the decompressed data.
+		snappyStartOffset := startOffset
+		snappyLength := length
+		length = compressedSize
+		startOffset = 0
+
+		// Validate the range.
+		if hrange != nil {
+			// For negative length we read everything.
+			if snappyLength < 0 {
+				snappyLength = actualSize - snappyStartOffset
+			}
+
+			// Reply back invalid range if the input offset and length fall out of range.
+			if snappyStartOffset > actualSize || snappyStartOffset+snappyLength > actualSize {
+				writeErrorResponse(w, ErrInvalidRange, r.URL)
+				return
+			}
+		}
+		objInfo.Size = snappyLength
+		setObjectHeaders(w, objInfo, hrange)
+		setHeadGetRespHeaders(w, r.URL.Query())
+
+		// Open a pipe for compression
+		// Where pipeWriter is actually passed to the getObject
+		pipeReader, pipeWriter := io.Pipe()
+		snappyReader := snappy.NewReader(pipeReader)
+
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			if hrange != nil {
+				// Discarding the data till the startOffset, if the range is enabled.
+				// goioutil.Discard imitates a dummy writer.
+				if _, pipeErr = io.CopyN(goioutil.Discard, snappyReader, snappyStartOffset); pipeErr != nil {
```

Comment:
From my understanding you're discarding all data until you reach the correct offset. This may have a huge performance impact on large objects - e.g. 5GB object and you try to access last 5 MB -> reading & discarding 99.9% of the object. I don't think we can afford this and we should try to do this differently.

Encryption had the same problem and we use a chunk-approach to avoid that performance impact.

Reply:
Hi @aead 
We eliminated the idea of having chunk-wise upload/read because it is infeasible to store chunk wise meta-data for all the chunks ( There can be many ) since the compressed size for a chunk will be un-predictable.
And there will be space/memory complexities here if we chose chunk wised upload/read for compression.
For eg, if there is a 5000MB (312 chunks with chunk size 16 MB) data uploaded . There should be 312 chunked compression meta data preserved. 
So finally we decided to discard since there was no other option here

Reply:
Why is it necessary to store k metadata entries (where `k = object size / chunk size`)?
Let's fix the chunk size to 16 MB and you upload e.g. 48 MB object. Now the `actual-size = 48MB` (trivial). Than you pass: 
1. 16 MB to `compress` which returns e.g 12 MB.   
2. 16 MB to `compress` which returns e.g 15 MB.
3. 16 MB to `compress` which returns e.g 10 MB.

=> The object size / compressed size = 37 MB

Now the user want to get the range 15 MB up to 20 MB. Than you have to load the chunks 1 and 2 (because 15 MB < 1*16 MB and 2*16 MB > 20 MB > 1*16MB). However you skip the first 15 MB of the first chunk and the last 12 MB of the second chunk. You don't touch the third chunk at all.

It's not necessary to read the exact byte range from the backend - you can skip and limit the packages after compressing. However with this approach your overhead is constant in worst case: 2*chunksize and not linear in object size.  

Or am I missing here something completely?

Reply:
There is a relation between actual chunk sizes and compressed chunk sizes which has to be stored.

Here in our example above , 

![image](https://user-images.githubusercontent.com/5410427/44004285-c3c56ce0-9e7d-11e8-828d-eb1aa492e6d6.png)

There will following difficulties in having chunked ranges
1) If the user wants to get the range (16-20) the snappy reader has to be informed to read the data from 12th mb till `x` , so here this info has to be preserved as meta data . 
2) In the above point , `x` is left unknown because there can be 2 compressed bytes decompressing to 4 actual bytes. So we can't set the offset/length on compressed data here, If so we will end up corrupting the data.
3) Snappy reader (or any compression reader) does not support `SEEK`

Reply:
My naive guess would be: `minCompressedSize(16 MB) = 786948`. The smaller `x` and the better snappy compresses the better is the approach I described with respect to read overhead. So maybe 64 KB for `x` may be reasonable size for read overhead (however I don't know the performance impact for snappy...)

Reply:
Yes i can understand your point here, But this cannot be done without the meta data info .
Let me take the following example,

![image](https://user-images.githubusercontent.com/5410427/44018125-674c0b92-9ef8-11e8-8666-bb15e820417e.png)

So there are 5 chunks here . The object's size is 59 and the actual size is 80. And let me leave `minCompressedSize(x)` for now.

If there is GET request with range , say (64-74), Now we know that it is enough to read the last chunk alone ( chunk 5 ) to serve the request. Here we should limit the reader with the compressed offset which will be 47 in this case. By this way reader omits the compressed data till 46 and read from 47th compressed byte. So it goes like

```
1) compressedChunkOffset = getCompressedChunkOffsetFromMetadata(metadata)
2) Set the reader offset with `compressedChunkOffset`
3) Read/Decompress it and write to the client
```
So we cant skip without knowing the compressed chunk's offset.
This applies even if we have `minCompressedSize()`, 

If there is a range request , say (50-70) , So we know that the requested range lies in chunk 4 and chunk 5. so, it is enough to read from `3*minCompressedSize(16)` till `5*16` until the range is satisfied.

So even here we need corresponding `compressedOffset` matching [`3*minCompressedSize(16)`]th actual byte . So that we can skip reading till `compressedOffset`.
We should not read from `1*16` till `5*16` for the range.

So here it goes like
```
1) ActualChunkRange = getActualChunkRange(request.Range)
// Which will return offset-->3*minCompressedSize(16) till-->5*16
2) CompressedChunkRange = getCompressedChunkRangeFromMetadata(ActualChunkRange)
// Which has compressed chunk offset
3) Limit the compressed reader to read from CompressedChunkRange.Offset (OR) Skip till CompressedChunkRange.Offset (which will read and discard)
4) Read the data from the reader and write to the client.
```
So here in both the cases we need compressed meta info stored. Which will be infeasible. And this will not be a problem if the compressed objects had fixed processed length (like encryption). The only problem here is non-predicted processed length. And also we cant set any constant metrics over compressed stream.

And secondly ,  Snappy returns a `Reader` for a `NewReader()` . Which cannot do a SEEK . AFAIK Only `Seeker` interface have Seek.

Reply:
@Praveenrajmani I still don't agree - (or total miss the point).
Before going into your example - do you agree on:
 - `compressedChunkSize(x) <= minCompressedSize(x)` holds for all `x`
 - `| compress(x) | <= | x |` where `|x|` is the length of `x` holds for all `x`

If that is true the following statement holds too:
`Let [a:b] be a range from a (inclusive) to b (exclusive) than:` 
<code>[compressedChunkSize(x)<sub>k</sub> : Σ<sub>k->n</sub> compressedChunkSize(x)] ⊆ [`k*minCompressedSize(x)` : `n*x`]</code>

This means the range `[k*minCompressedSize(x) : n*x]` **always** includes the requested compressed range - it may be larger but there exists two offsets `i` and `j` within the range such that `decompress([i:j]` is exactly our requested data. However due to compression we don't know `i` and `j` without additional metadata (That's your point and I agree on that). I clam we don't need to know `i` and `j` because we can decompress the entire range first. Than we can set `i'` to client offset and `j'` to client length and write the range `[i' : j']` to the client. Also I clam that is is significatly more efficient than an entire read from the beginning. For my argumentation `Seek` functionality is not required.   

### Your example - assuming 5 chunk object as in table:
- Range from 64-74:
   1. Chunk-Size = 16 => 64 / 16 = 4 and 5*16 > 74
   2. Read from backend: `4*minCompressedSize(16) - 5*16` (or since 5*16 > size til end)
   3. Decompress `4*minCompressedSize(16) - 5*16` returns a range [i:j]
   4. Write [j-10:j] to client.

 That should be the correct range - notice I didn't looked at the size of a chunk at all.

- Range from 50-70:
  1. Chunk-Size = 16 => 50 / 16 = 3 and 5*16 > 70
  2. Read from backend: `3*minCompressedSize(16) - 5*16` (or since 5*16 > size til end)
  3. Decompress `3*minCompressedSize(16) - 5*16` returns a range [i:j]
  4. Write [j-24:j-4] to client.

That should be the correct range.

The whole point of `minCompressedSize` is to eliminate the issue of storing the compressed chunk size as metadata. We **don't** compute the correct compressed offsets - we only want a range from the backend which guarantees that the range we're interested in is part of it. As soon as we can achieve that can **always** discard the beginning and end of our range to extract only the range we're interested in.

Reply:
Also an performance aspect: The closer the actual size of `compressedSizeOf(chunk)` is to `minCompressedSize(|chunk|) the less the overhead we get with the scheme. Ideally if we had an ideal compression function which process compressed out such that `compressed `compressedSizeOf(chunk) == minCompressedSize(|chunk|)` for any `chunk` the overhead would be 0.